### ISSUE TITLE
pylab image_masked is broken

### ISSUE THREAD

**Issue by efiring** [matplotlib/matplotlib#6509]
It looks like the input values are getting clipped instead of registering as over and under.

With v2.x:
![image_masked_v2](https://cloud.githubusercontent.com/assets/85125/15666726/134b9824-26ae-11e6-84ed-02fbb4b731f6.png)

It should look like this, from v1.5:
![image_masked_v1 5](https://cloud.githubusercontent.com/assets/85125/15666787/61d4044a-26ae-11e6-8e92-50480ca2fbd3.png)

**Comment by efiring** [comment]
attn: @mdboom
For the first plot, the problem is that support for over and under values has been lost when image interpolation is anything other than 'none' or 'nearest'.
```python
cmap = get_cmap('viridis')
cmap.set_over('c')
cmap.set_under('r')
imshow(randn(20,20), vmin=-0.5, vmax=0.5, cmap=cmap, interpolation='bilinear')
imshow(randn(20,20), vmin=-0.5, vmax=0.5, cmap=cmap, interpolation='none')
imshow(randn(20,20), vmin=-0.5, vmax=0.5, cmap=cmap, interpolation='nearest')
```

**Comment by efiring** [comment]
For the second plot, BoundaryNorm is not broken, so it looks like the problem is that imshow is no longer recognizing the integer indexing into the colormap LUT.
```python
from matplotlib.colors import BoundaryNorm
bn = BoundaryNorm([-1, -0.5, -0.2, 0, 0.2, 0.5, 1], ncolors=256, clip=False)
xx = np.linspace(-2, 2, 100).reshape(10, 10)
fig, (ax0, ax1) = subplots(ncols=2)
ax0.pcolormesh(xx, norm=bn, cmap=cmap)
ax1.imshow(xx, norm=bn, cmap=cmap)
```

**Comment by tacaswell** [comment]
It seems that `_image.resample` is clipping everything to [0, 1], independent of input type (which is why we moved the normalization first to begin with).  This causes 2 problems.  The first is that the `BoundaryNorm` mappings all end up being 1 or 0 and second is it cuts off the over/under.  I have made some progress on the second issue.

![figure_1](https://cloud.githubusercontent.com/assets/199813/16029412/d2c37d08-31b6-11e6-87b4-f0796eb4d178.png)

which is generated off of the commit in #6581

**Comment by mdboom** [comment]
I think @tacaswell's solution in #6581 is actually pretty elegant and doesn't add much runtime overhead.

**Comment by tacaswell** [comment]
It still completely fails on BoundaryNorm and NoNorm use cases though.

**Comment by mdboom** [comment]
Did you see my comment in #6581 about a fix for BoundaryNorm?

**Comment by efiring** [comment]
@mdboom, does image handling always go through `_ImageBase._make_image()`?  If so, I think the integer indexing case can be handled by doing the colormapping immediately when the norm returns integers, and then proceeding as if an RGBA array had been supplied from the start.  This seems to me to be in the spirit of the integer indexing.
I can try implementing this later today, unless someone else would like to do so, or a flaw in the strategy becomes clear.

**Comment by mdboom** [comment]
> If so, I think the integer indexing case can be handled by doing the colormapping immediately when the norm returns integers, and then proceeding as if an RGBA array had been supplied from the start. This seems to me to be in the spirit of the integer indexing.

Doesn't that leaves us right back where we started?  Interpolating in RGBA space and all of the incorrect results that that causes?

**Comment by efiring** [comment]
@mdboom, only in the case of integer colormap indexing, where interpolation other than none or nearest is not really appropriate anyway--so I don't see any real problem. It would be the same as the present case with RGBA input, right?

**Comment by mdboom** [comment]
I see your point.  I'm sure there are some users that pass in integers representing continuous data anyway.

Seems like a reasonable compromise -- if ints passed in, map to rgba first.  If floats, interpolate then color map.

**Comment by QuLogic** [comment]
I believe this was fixed by #6581.
